ocrd_models/ocrd_models/ocrd_mets.py

```diff
@@ -284,13 +410,21 @@ def remove_file_group(self, USE, recursive=False, force=False):
             if not recursive:
                 raise Exception("fileGrp %s is not empty and recursive wasn't set" % USE)
             for f in files:
+                # NOTE: Here we know the fileGrp, we should pass it as a parameter
```

Comment:
OK, so because

```python
if self._cache_flag:                                                                                                                                       
    if fileGrp:                                                                                                                                            
        if isinstance(fileGrp, str):                                                                                                                       
            candidates += self._file_cache.get(fileGrp, {}).values()                                                                                       
        else:                                                                                                                                              
            candidates = [x for fileGrp_needle, el_file_list in self._file_cache.items() if fileGrp.match(fileGrp_needle) for x in el_file_list.values()]  
    else:                                                                                                                                                  
        candidates = [el_file for id_to_file in self._file_cache.values() for el_file in id_to_file.values()]                                              
else:                                                                                                                                                      
    candidates = self._tree.getroot().xpath('//mets:file', namespaces=NS)                                                                                  
```

could use the fileGrp for accessing the cache, fine with me.


But thinking about the code here, we could at least use `[@ID=%s'] % ID` if the ID was provided for the non-caching version, that would reduce the number of candidates to, ideally, 1.

Reply:
Without caching we still end up with a list of `mets:file` elements after 
`candidates = self._tree.getroot().xpath('//mets:file', namespaces=NS)`.
I don't see how the number of candidates can be 1.

In the best case, we match the first element and leave. 
In the worst case, we iterate the entire list to get the last element.
So the average is N/2.

Reply:
Oh, okay I get what you meant. But still we end up with a linear time. Unless you provide the exact tree path to the element tree, the element tree still ends up iterating a list of `mets:file`. We then get only 1 candidate, true and match just that candidate in the loop afterward.